🌱 logging: adjust reconcilers to log object owners

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
With this PR we're logging all relevant k/v pairs for all our resources.

This PR also contains 1-2 smaller other logging fixes and an improvement to our Tiltfile so the ServiceAccount of Grafana is **always** deployed before the Grafana deployment.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

sbueringer commented on util/log/log.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+// AddObjectHierarchy adds the hierarchy of an object as k/v pairs to the logger in ctx.
+func AddObjectHierarchy(ctx context.Context, c client.Client, obj metav1.Object) (context.Context, logr.Logger, error) {
```
Would be good to get some feedback about that func first.

I would then:
* add corresponding unit tests
* document it in the v1.2=>v1.3 doc that providers can use it too

I intentionally:
* didn't resolve the entire ownerRef tree (not sure how deep they are in folks using CAPI in non-mainstream ways)
* did resolve MachineSets as that's a known case and very useful to include MachineDeployments in Machine logs
* did use the typed client to get the MachineSet as the MachineSet will be automatically cached, it would have been a lot more complicated to require an unstructured cached client
* didn't include an allow-list of ownerRefs we want to add as k/v pairs. We could do that but I'm not sure if it's okay to hard-code "KubeadmControlPlane" (things like "MachineDeployment", "MachineSet" should be fine.

This change gives us the following k/v pairs:

```md
* bootstrap
  * KubeadmConfig
    * kv: Cluster | MachinePool | KubeadmConfig
    * kv: Cluster | KubeadmControlPlane | Machine | KubeadmConfig
    * kv: Cluster | MachineDeployment | MachineSet | Machine | KubeadmConfig
* controlplane
  * KubeadmControlPlane
    * kv: Cluster | KubeadmControlPlane
* core
  * Cluster
    * kv: Cluster 
  * MachineDeployment
    * kv: Cluster | MachineDeployment
  * MachineSet
    * kv: Cluster | MachineDeployment | MachineSet
  * Machine
    * kv: Cluster | MachineDeployment | MachineSet | Machine
    * kv: Cluster | KubeadmControlPlane | Machine
  * MachineHealthcheck
    * kv: Cluster | MachineHealthcheck

* CAPD
  * DockerCluster
    * kv: Cluster | DockerCluster
  * DockerMachine
    * kv: Cluster | MachineDeployment | MachineSet | Machine | DockerMachine
    * kv: Cluster | KubeadmControlPlane | Machine | DockerMachine 
  * DockerMachinePool
    * kv: Cluster | MachinePool | DockerMachinePool
```

sbueringer replied:
cc @fabriziopandini

fabriziopandini replied:
The result is what we expect, so I'm +1 for the change with the current implementation (we can eventually improve later)
The only improvement that comes to my mind is to avoid to add twice the same owner ref, that is something that could happen e.g when the topology controllers adds additional owner refs to ensure object cleanup

sbueringer replied:
This is safe from the logger-side. The log k/v pairs are deduplicated by `key`

sbueringer replied:
I'll add it here anyway. Probably better to deduplicate here, vs otherwise it will be done on every log call

fabriziopandini commented on internal/controllers/machine/machine_controller_phases.go:
```diff
@@ -98,7 +98,7 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
 	if err != nil {
 		if apierrors.IsNotFound(errors.Cause(err)) {
-			log.Info("could not find external ref, requeueing", "refGVK", ref.GroupVersionKind(), "refName", ref.Name, "Machine", klog.KObj(m))
+			log.Info("could not find external ref, requeueing", "refGVK", ref.GroupVersionKind(), "refName", ref.Name)
```
Can we log ref.Kind (key) ref.Namespace/name (value) so it is consistent with other key values pairs

sbueringer replied:
Done

fabriziopandini commented on internal/controllers/machineset/machineset_controller.go:
```diff
@@ -131,6 +130,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
+	ctx, log, err := clog.AddObjectHierarchy(ctx, r.Client, machineSet)
```
Q: if we add this, should we drop e.g L138 (previous code adding object hierarchy)

sbueringer replied:
The Cluster is actually not in the ownerRefs only the MachineDeployment. 
> https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1565639347489738752/artifacts/clusters/bootstrap/resources/quick-start-drwwuo/MachineSet/quick-start-yw49at-md-0-98fqk-6785587f97.yaml

If there are cases where it would be the logger automatically deduplicates duplicate keys

fabriziopandini replied:
But the Cluster will be added by AddObjectHierarchy, is that right?

sbueringer replied:
If the object we pass in has the Cluster as an ownerRef => yes

fabriziopandini commented on util/log/log.go:
```diff
@@ -0,0 +1,97 @@
+			return nil, errors.Wrapf(err, "failed to get owners: failed to get MachineSet %s", klog.KRef(obj.GetNamespace(), ownerRef.Name))
+		}
+
+		for _, ref := range ms.GetOwnerReferences() {
+			owners = append(owners, owner{
+				Kind:      ref.Kind,
+				Namespace: obj.GetNamespace(),
+				Name:      ref.Name,
+			})
+		}
```
Q: Could this be replaced by 
```suggestion
		owners = append(owners, getOwners(ctx, c, ms)...)
```

sbueringer replied:
We could. Although I also have to handle the error that is returned. So the resulting code wouldn't be much shorter.

Wonder if we should keep the current version given that it's a bit simpler to understand than the recursion.
(even if we call it recursively a MS never has another MS as ownerRef so it is technically equivalent to the other code, but it's harder to see that)

fabriziopandini replied:
No problem, it is ok as it is

sbueringer commented:
Puhsed unit tests and documented the new fund in the migration doc.

sbueringer commented:
Updated the note in the migration doc, renamed the fund and added godoc. (I dropped the func from DockerMachinePool as I noticed that we don't need it there)

fabriziopandini commented:
/lgtm
/approve
